File: src/test/java/com/raychenon/leetcode/RottingOrangesTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 12/2/2020
 * https://leetcode.com/problems/rotting-oranges/
 */
public class RottingOrangesTest {

    private static RottingOranges instance;

    @BeforeClass
    public static void setUp() {
        instance = new RottingOranges();
    }

    @Test
    public void example1Test() {
        int[][] matrix = new int[][]{
                {2, 1, 1},
                {1, 1, 0},
                {0, 1, 1}
        };

        assertOrangesRotting(4, matrix);
    }

    @Test
    public void example2Test() {
        int[][] matrix = new int[][]{
                {2, 1, 1},
                {0, 1, 1},
                {1, 0, 1}
        };

        assertOrangesRotting(-1, matrix);
    }

    @Test
    public void example3Test() {
        int[][] matrix = new int[][]{
                {0, 2}
        };

        assertOrangesRotting(0, matrix);
    }

    private void assertOrangesRotting(int minutesExpected, int[][] grid) {
        assertEquals(minutesExpected, instance.orangesRottingBFS(grid));
        // the 2nd method modifies in space the grid
        assertEquals(minutesExpected, instance.orangesRotting(grid));
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/RottingOrangesTest.kt
package com.raychenon.kotlin.leetcode

import junit.framework.TestCase
import org.junit.Test

/**
 * User: raychenon
 * Date: 16/2/2020
 * https://leetcode.com/problems/rotting-oranges/
 */
class RottingOrangesTest {

    @Test
    fun example1Test() {
        val matrix =
            arrayOf(intArrayOf(2, 1, 1), intArrayOf(1, 1, 0), intArrayOf(0, 1, 1))
        evaluateRottingOranges(4, matrix)
    }

    @Test
    fun example2Test() {
        val matrix =
            arrayOf(intArrayOf(2, 1, 1), intArrayOf(0, 1, 1), intArrayOf(1, 0, 1))
        evaluateRottingOranges(-1, matrix)
    }

    @Test
    fun example3Test() {
        val matrix = arrayOf(intArrayOf(0, 2))
        evaluateRottingOranges(0, matrix)
    }

    fun evaluateRottingOranges(expectedMinutes: Int, grid: Array<IntArray>): Unit {
        // does not modify the grid
        TestCase.assertEquals(expectedMinutes, RottingOranges.orangesRottingBFS(grid))
        // modifies the grid in-place
        TestCase.assertEquals(expectedMinutes, RottingOranges.orangesRotting(grid))
    }

}